# The toytext environment chosen for this assignment is <i> Frozen Lake </i> 
## The episode ends when you reach the goal or fall in a hole. You receive a reward of 1 if you reach the goal, and zero otherwise. (This is fixed!)
## NOTE: The ice is slippery, so you won't always move in the direction you intend.
 <img src="/ipynb.images/frozen_lake.png">

First we will install the necessary libraries and establish a baseline model

In [1]:
import numpy as np
import gym
import random
import time
from IPython.display import clear_output

In [2]:
import gym # loading the Gym library
 
env = gym.make("FrozenLake-v0")
env.reset()                    
env.render()


SFFF
FHFH
FFFH
HFFG


In [3]:
#Checking the action and state size
action_size = env.action_space.n
print("Action:", action_size)
state_size = env.observation_space.n
print("State:", state_size)

Action: 4
State: 16


In [4]:
#initialize q table
qtable = np.zeros((state_size,action_size))
print(qtable)

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]


In [5]:
#hyperparameters - Baseline Model
#Next step: QLearning 

total_episodes = 5000        # Total episodes
# total_test_episodes = 100     # Total test episodes
max_steps = 99                # Max steps per episode

learning_rate = 0.7           # Learning rate
gamma = 0.8                 # Discounting rate

# Exploration parameters
epsilon = 1.0                 # Exploration rate
max_epsilon = 1.0             # Exploration probability at start
min_epsilon = 0.01            # Minimum exploration probability 
decay_rate = 0.001             # Exponential decay rate for exploration prob

In [6]:
#Q Learning 

rewards_all=[]


for episode in range(total_episodes):
    state=env.reset()
    step=0
    done=False
    reward_curr=0
    
    for step in range(max_steps):
        # 3. Choose an action a in the current world state (s)
        ## First we randomize a number
        exp_exp_tradeoff = random.uniform(0,1)
        
        ## If this number > greater than epsilon --> exploitation (taking the biggest Q value for this state)
        if exp_exp_tradeoff > epsilon:
            action = np.argmax(qtable[state,:]) #using max Policy
            
        # Else doing a random choice --> exploration
        else:
            action = env.action_space.sample()
        
        # Take the action (a) and observe the outcome state(s') and reward (r)
        new_state, reward, done, info = env.step(action)
        
        # Update Q(s,a):= Q(s,a) + lr [R(s,a) + gamma * max Q(s',a') - Q(s,a)]
        qtable[state, action] = qtable[state, action] + learning_rate * (reward + gamma * 
                                    np.max(qtable[new_state, :]) - qtable[state, action]) #using max policy
        
        
        state = new_state
        reward_curr+=reward
        
        # If done : finish episode
        if done == True: 
            break
            
    # Reduce epsilon (because we need less and less exploration)
    epsilon = min_epsilon + (max_epsilon - min_epsilon)*np.exp(-decay_rate*episode)
    rewards_all.append(reward_curr)
    
    

In [7]:
#Calculate and print the average reward per thousand episodes

rewards_per_thousand = np.split(np.array(rewards_all), total_episodes/1000)
count=1000
print("Average reward per thousand episodes:")

for r in rewards_per_thousand:
    print(count , " : " , str(sum(r/1000)))
    count+=1000

Average reward per thousand episodes:
1000  :  0.023000000000000013
2000  :  0.07300000000000005
3000  :  0.1290000000000001
4000  :  0.20000000000000015
5000  :  0.23900000000000018


Highest average rewards = 0.239

In [8]:
# #hyperparameters - Trial & Error

# total_episodes = 10000        # Total episodes
# # total_test_episodes = 100     # Total test episodes
# max_steps = 99                # Max steps per episode

# learning_rate = 0.9           # Learning rate
# gamma = 0.9                 # Discounting rate

# # Exploration parameters
# epsilon = 1.0                 # Exploration rate
# max_epsilon = 1.0             # Exploration probability at start
# min_epsilon = 0.001            # Minimum exploration probability 
# decay_rate = 0.001             # Exponential decay rate for exploration prob

In [29]:
#hyperparameters - Manually Optimized Model

total_episodes = 10000        # Total episodes
# total_test_episodes = 100     # Total test episodes
max_steps = 100                # Max steps per episode

learning_rate = 0.1           # Learning rate
gamma = 0.99                 # Discounting rate

# Exploration parameters
epsilon = 1                 # Exploration rate
max_epsilon = 1.0             # Exploration probability at start
min_epsilon = 0.001            # Minimum exploration probability 
decay_rate = 0.001             # Exponential decay rate for exploration prob

In [30]:
#Q Learning 

rewards_all=[]


for episode in range(total_episodes):
    state=env.reset()
    step=0
    done=False
    reward_curr=0
    
    for step in range(max_steps):
        # 3. Choose an action a in the current world state (s)
        ## First we randomize a number
        exp_exp_tradeoff = random.uniform(0,1)
        
        ## If this number > greater than epsilon --> exploitation (taking the biggest Q value for this state)
        if exp_exp_tradeoff > epsilon:
            action = np.argmax(qtable[state,:]) #using max Policy
                
        # Else doing a random choice --> exploration
        else:
            action = env.action_space.sample()
        
        # Take the action (a) and observe the outcome state(s') and reward (r)
        new_state, reward, done, info = env.step(action)
        
        # Update Q(s,a):= Q(s,a) + lr [R(s,a) + gamma * max Q(s',a') - Q(s,a)]
        qtable[state, action] = qtable[state, action] + learning_rate * (reward + gamma * 
                                    np.max(qtable[new_state, :]) - qtable[state, action]) #using max policy
        
        
        state = new_state
        reward_curr+=reward
        
        # If done : finish episode
        if done == True: 
            break
            
    # Reduce epsilon (because we need less and less exploration)
    epsilon = min_epsilon + (max_epsilon - min_epsilon)*np.exp(-decay_rate*episode)
    rewards_all.append(reward_curr)
    
    

In [31]:
len(rewards_all)

10000

In [32]:
#Calculate and print the average reward per thousand episodes

rewards_per_thousand = np.split(np.array(rewards_all), total_episodes/1000)
count=1000
print("Average reward per thousand episodes:")

for r in rewards_per_thousand:
    print(count , " : " , str(sum(r/1000)))
    count+=1000

Average reward per thousand episodes:
1000  :  0.05100000000000004
2000  :  0.21100000000000016
3000  :  0.4150000000000003
4000  :  0.5880000000000004
5000  :  0.6620000000000005
6000  :  0.7070000000000005
7000  :  0.7120000000000005
8000  :  0.7420000000000005
9000  :  0.7470000000000006
10000  :  0.7290000000000005


# Highest avg reward = 0.747
# Which is great considering our baseline

Hyperparameter Tuning Observation:
--> Increasing number of episodes improved the avg reward somewhat but alone did not make reward high <br>
--> Increasing max_steps did not have much effect on the rewards<br>
--> Increasing learning rate alone improved the rewards but performance varied when other paramters were changed at the same time<br>
--> Reducing gamma reduced the rewards significantly and increasing gamma increased the rewards<br>
--> Reducing exploration rate did improve rewards somewhat but after multiple experiments it came to notice that the highest amount a reward could go upto was limited(only around 0.4 something) and nowhere near 1.0<br>
--> Increasing decay rate had similar effect as that of exploration rate reduction<br>
--> Reducing the minimum exploration probability improved the rewards and increased the time by a little bit for the agent<br>

*Note: Only Max Value Policy is used to adjust the hyperparameters<br>


# Now we will change the policy from maximum to random and observe the results
Another policy - Minimum Value policy has been commented alongside this code and can be uncommented for experiment.
Results for both are mentioned later 

In [33]:
#Q Learning 

rewards_all=[]


for episode in range(total_episodes):
    state=env.reset()
    step=0
    done=False
    reward_curr=0
    
    for step in range(max_steps):
        # 3. Choose an action a in the current world state (s)
        ## First we randomize a number
        exp_exp_tradeoff = random.uniform(0,1)
        
        ## If this number > greater than epsilon --> exploitation (taking the biggest Q value for this state)
        if exp_exp_tradeoff > epsilon:
#             action = np.argmax(qtable[state,:]) #using max Policy
#             action = np.argmin(qtable[state,:]) #using min Policy
            action = np.random.choice(qtable[state,:].size) # using random Policy
                
        # Else doing a random choice --> exploration
        else:
            action = env.action_space.sample()
        
        # Take the action (a) and observe the outcome state(s') and reward (r)
        new_state, reward, done, info = env.step(action)
        
        # Update Q(s,a):= Q(s,a) + lr [R(s,a) + gamma * max Q(s',a') - Q(s,a)]
#         qtable[state, action] = qtable[state, action] + learning_rate * (reward + gamma * 
#                                     np.max(qtable[new_state, :]) - qtable[state, action]) #using max policy
#         qtable[state, action] = qtable[state, action] + learning_rate * (reward + gamma * 
#                                     np.min(qtable[new_state, :]) - qtable[state, action]) #using max policy
        qtable[state, action] = qtable[state, action] + learning_rate * (reward + gamma * 
                                    np.random.choice(qtable[new_state, :]) - qtable[state, action]) #using max policy
        
        
        state = new_state
        reward_curr+=reward
        
        # If done : finish episode
        if done == True: 
            break
            
    # Reduce epsilon (because we need less and less exploration)
    epsilon = min_epsilon + (max_epsilon - min_epsilon)*np.exp(-decay_rate*episode)
    rewards_all.append(reward_curr)
    
    

In [34]:
#Calculate and print the average reward per thousand episodes

rewards_per_thousand = np.split(np.array(rewards_all), total_episodes/1000)
count=1000
print("Average reward per thousand episodes:")

for r in rewards_per_thousand:
    print(count , " : " , str(sum(r/1000)))
    count+=1000

Average reward per thousand episodes:
1000  :  0.009000000000000001
2000  :  0.008
3000  :  0.014000000000000005
4000  :  0.013000000000000005
5000  :  0.009000000000000001
6000  :  0.02000000000000001
7000  :  0.016000000000000007
8000  :  0.014000000000000005
9000  :  0.017000000000000008
10000  :  0.016000000000000007


The results are pretty random as we can see above. 
# The rewards extremely low which means that we will rarely almost never reach the goal with this policy. We can run the same below and verify this. 

In [35]:
#Run to see agent play FrozenLake (End result of each episode with total number of steps)
env.reset()
rewards_all =[]
total_step=0
step_count=0

for episode in range(5): # can go upto total_episodes
    state=env.reset()
    done=False
    print("****Episode", episode+1, "****")
    time.sleep(1)
    
    for step in range(max_steps):
        
        # Take the action (index) that have the maximum expected future reward given that state
        action = np.argmax(qtable[state,:])
        
        new_state, reward, done, info = env.step(action)
        
        if step==max_steps:
            print(epsilon)
        
        if done:
            clear_output(wait=True)
            env.render()
            if reward==1:
                print("*** REACHED THE GOAL! *** ", step, "Steps")
                time.sleep(3)
            else:
                print("*** FELL THROUGH A HOLE! ***", step, "Steps")
                time.sleep(3)
            clear_output(wait=True)
            break
        
        state = new_state
        total_step+=step
        step_count+=1
    

print("Average steps:", total_step/step_count)
env.close()

Average steps: 11.35576923076923


# From the above piece of code, we see that Q Learning uses value based iteration, keeping the policy fixed. At a time, we can have one policy only that defines what actions to take based on a state. A policy is a function or a set of rules. 


The expected lifetime value in the Bellman Equation is the duration for which the agent takes action in the environment.

** The total steps almost never reaches the max_steps

Important Citations:
https://www.freecodecamp.org/news/diving-deeper-into-reinforcement-learning-with-q-learning-c18d0db58efe/
https://www.freecodecamp.org/news/an-introduction-to-reinforcement-learning-4339519de419/
https://www.youtube.com/watch?v=Hj4cCmC7jv4

Copyright 2020 Kartik Kumar

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.